taskmanager/server.go

```diff
@@ -0,0 +1,155 @@
+package taskmanager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/tcsc/levity/api"
+	"github.com/tcsc/levity/registry"
+	"github.com/tcsc/levity/task"
+)
+
+// Error is an error code describing things that can go wrong with the
+// TaskManager API service.
+type Error int
+
+const (
+	// NoSuchTask indicates that the operation was targeted at a task
+	// that does not exist.
+	NoSuchTask Error = 1
+)
+
+func (e Error) Error() string {
+	switch e {
+	case NoSuchTask:
+		return "No such task"
+	}
+	return fmt.Sprintf("Unknown error code: %d", int(e))
+}
+
+// TaskRegistry abstracts out the desinition of the task0id-to-task mapping, so
+// we can mock it ut for testing
+type TaskRegistry interface {
+	Register(ctx context.Context, t *task.Task) (string, error)
+	Lookup(id string) *task.Task
+}
+
+// Server is an implementation of the TaskManager API
+type Server struct {
+	api.UnimplementedTaskManagerServer
+	registry TaskRegistry
+}
+
+// New creates and initialises a new Server with default settings
+func New() *Server {
+	return withRegistry(registry.New())
+}
+
+// withRegistry creates and initialises a new Server with a custom task registry
+// implementation
+func withRegistry(registry TaskRegistry) *Server {
+	return &Server{registry: registry}
+}
+
+// StartTask attempts to start and register a task with the task manager.
+func (server *Server) StartTask(ctx context.Context, req *api.StartTaskRequest) (*api.StartTaskResponse, error) {
+	t := task.New(
+		req.GetBinary(),
+		req.GetWorkingDir(),
+		req.GetEnvironment(),
+		req.GetArgs()...)
+
+	// Start the task *before* registering it, as we don't want to expose the task
```

Comment:
I think this may expose the client to unexpected side-effects, for example what if the task was started but failed to register? It will basically be left orphaned with the client having no ability to know this and/or stop it. Is there a way to handle this scenario better?

Reply:
The task gets killed if registration fails a few lines down.

The other alternative I thought of was essentially registering a `nil` task to allocate a unique ID. Once we have that, simply assigning anew task to the empty slot cannot fail, so the flow would look something like this:

1. Get a unique task ID, which could potentially fail
2. On success, create the task, which might fail
3. On success, bind the task to the Id, which cannot fail
4. Fail all RPC calls that return a `nil` task with `no-such-task`

That feels a bit better to me.